src/java.net.http/share/classes/java/net/http/HttpResponse.java

```diff
@@ -748,6 +749,34 @@ public static <T> BodyHandler<T> buffering(BodyHandler<T> downstreamHandler,
                      .buffering(downstreamHandler.apply(responseInfo),
                                 bufferSize);
          }
+
+        /**
+         * {@return a {@code BodyHandler} limiting the number of bytes consumed
+         * and passed to the given downstream {@code BodyHandler}}
+         * <p>
+         * If the number of bytes received exceeds the maximum number of bytes
+         * desired as indicated by the given {@code capacity},
+         * {@link BodySubscriber#onError(Throwable) onError} is called on the
+         * downstream {@code BodySubscriber} with an {@link IOException}
+         * indicating that the capacity is exceeded, and the upstream
+         * subscription is cancelled.
+         *
+         * @param downstreamHandler the downstream handler to pass received data to
+         * @param capacity the maximum number of bytes that are allowed
+         * @throws IllegalArgumentException if {@code capacity < 0}
+         * @since 25
+         */
+        public static <T> BodyHandler<T> limiting(BodyHandler<T> downstreamHandler, long capacity) {
+            Objects.requireNonNull(downstreamHandler, "downstreamHandler");
+            if (capacity < 0) {
+                throw new IllegalArgumentException("was expecting \"capacity >= 0\", found: " + capacity);
```

Comment:
Nit - in some other areas, we use messages of the form `("capacity must not be negative: " + capacity)`. The current proposed message is more commonly used in test case exceptions, so it feels a bit odd to be seeing it in this form.

Reply:
Fixed in 23f79a20476f784058801454a28d37f37df92982.